lsp-types/src/Language/LSP/Types/Uri.hs

```diff
@@ -180,16 +178,16 @@ instance Binary NormalizedFilePath where
   put (NormalizedFilePath _ fp) = put fp
   get = do
     v <- Data.Binary.get :: Get ShortByteString
-    case decodeFilePath v of
-      Left e -> fail (show e)
-      Right v' ->
-        return (NormalizedFilePath (internalNormalizedFilePathToUri v') v)
+    let v' = decodeFilePath v
+    return (NormalizedFilePath (internalNormalizedFilePathToUri v') v)
 
-encodeFilePath :: String -> ShortByteString
-encodeFilePath = BS.toShort . T.encodeUtf8 . T.pack
+-- | Convert 'FilePath' to a UTF-8 encoded 'ShortByteString'
+encodeFilePath :: FilePath -> ShortByteString
+encodeFilePath = BS.pack . UTF8.encode
```

Comment:
My local benchmark results show that this takes more time than `BS.toShort . T.encodeUtf8 . T.pack` in most cases. Quite surprising. Waiting for benchmark results on CI. https://github.com/haskell/haskell-language-server/pull/3067

Reply:
Ok, maybe not that surprising given that `text` is likely much more optimised than `utf8-string`.

The best option here would be to leverage the fact that `text-2.0` is internally isomorphic to a `ShortByteString` plus a pair of ints for slicing, and encoded in UTF-8. So you could just do:

```
#if min_version_text(2,0)
import Data.ByteString.Short
import qualified Data.Text as Text
import Data.Text.Array(Array(..))
import Data.Text.Internal (Text(Text))
...

encodeFilePath :: FilePath -> ShortByteString 
encodeFilePath fp = case Text.pack fp of
  Text (ByteArray sbs) _ _ -> SBS sbs

decodeFilePath :: ShortByteString -> FilePath
decodeFilePath (SBS fp) = Text.unpack $ Text (ByteArray fp) 0 (length fp) 
```